docs/contributing/api-mocking.md

```diff
@@ -1,53 +1,176 @@
+However, when `Diagnostics` recovers a branch from an error, the internals of
+that computation are discarded.  This means that any state tracked in that
+branch is discarded and so the mock would have no way to know what calls were
+made and which expectations to exhaust.  This would mean that we could not
+create a situation where a request fails, the failure is recovered from and then
+the request retries and succeeds because the mock would never remember if it had
+made the failing call or not.
 
-Expectations are checked and matched in the order in which they are added.
+`MockApi` then has to be below `Diagnostics` in the carrier stack.  This allows
+it to keep track of its state regardless of what errors are thrown, caught or
+recovered from.
 
-Any unsatisfied expectations at the end of the test will trigger an assertion error.  Remember that the requests must match exactly (via `Eq`).
+### Expectations
 
-Examples:
+The core of the mocking is `ApiExpectations`.  `ApiExpectations` are both a
+description of behavior and an assertion about what behavior should occur in the
+test.  The requests that the expectations match are of type `FossaApiClientF a`,
+where the return-type, `a`, can be different depending on the type of request.
+This leads to an awkward situation of matching return types to the request types
+without knowing the actual type of the value.
+
+The expectations need to ensure that request and return types match, but
+otherwise can "forget" the type.  This can be achieved with a GADT to "erase"
+the return type.
 
 ```haskell
--- return the same value every time a call is made.
-GetOrganization `alwaysReturns` Fixtures.organization
+data ApiExpectation where
+  ApiExpectation ::
+    ExpectationRepetition
+    -> FossaApiClientF a
+    -> ApiResult a
+    -> ApiExpectation
+```
 
--- Simulate an error
-GetProject revision `fails` "Mock HTTP error"
+Expectations consist of:
+1. `ExpectationRepetition`: a repetition specification used to track when it is
+   exhausted
+1. `FossaApiClientF a`: a request to match against
+1. `ApiResult a`: a behavior such as returning a value or raising an error
 
--- Return one value the first time, then another on any subsequent calls
-(GetScan locator scanId)
-  `returnsOnce` Fixtures.scanResponse{responseScanStatus = Nothing}
-(GetScan locator scanId)
-  `alwaysReturns` Fixtures.scanResponse{responseScanStatus = Just "Available"}
-```
+The repetition specifies how to exhaust the the expectation.  It's currently
+only two options, but there is room for expansion.
+
+- `Once` means the expectation should be exhausted after one match.
+- `Always` means the expectation should never be exhausted.
+
+The request is used to match incoming requests with `Eq`.  It would be more
+flexible as a predicate, but this would make error messages harder to write.  A
+copy of the request is very useful in debugging and writing errors.
```

Comment:
It may be worth coming up with some error-friendly comparison typeclass based on `Eq`, but that's a later improvement if we need it..

Reply:
Yeah, I'd really like it to figure this out automatically, which might mean we need a new typeclass (hopefully not TH) or to use Typeable.

Reply:
Yeah this is non-trivial, but if we need better equality comparison in general, then a `StructuralDiff` typeclass might be necessary.